From b5056d265ccef037ffb2fc3a2f620ebfd787a9ac Mon Sep 17 00:00:00 2001
From: Chen Zhongjin <chenzhongjin@huawei.com>
Date: Mon, 29 May 2023 17:21:07 +0800
Subject: [PATCH 208/263] [Huawei] samples: bmc: Close temporary fd in bmctool

Offering: HULK
hulk inclusion
category: bugfix
bugzilla: N/A

--------------------------------

In bmctool the cgroup_fd and prog_fd for prog load/unload are not closed
after used, which can cause fd leak.
Close temporary fd after use it.

Signed-off-by: Chen Zhongjin <chenzhongjin@huawei.com>
---
 samples/bpf/bmc/tool.c | 25 ++++++++++++++++---------
 1 file changed, 16 insertions(+), 9 deletions(-)

diff --git a/samples/bpf/bmc/tool.c b/samples/bpf/bmc/tool.c
index c88bba396f50..1fe4cba0ce14 100644
--- a/samples/bpf/bmc/tool.c
+++ b/samples/bpf/bmc/tool.c
@@ -493,6 +493,7 @@ static int setup_bpf(void)
 	if (add_tail_call())
 		goto error_add_port;
 
+	close(bmc.cgroup_fd);
 	return 0;
 
 error_add_port:
@@ -669,10 +670,10 @@ static int do_prog_load(int argc, char *argv[])
 static int do_prog_unload(int argc, char *argv[])
 {
 	int i;
-	int err;
-	int prog_fd;
-	int cgroup_fd;
-	int map_fd;
+	int err = -1;
+	int prog_fd = -1;
+	int cgroup_fd = -1;
+	int map_fd = -1;
 	char *interface_map_path = NULL;
 	__u32 ifindex;
 	__u32 key;
@@ -684,7 +685,7 @@ static int do_prog_unload(int argc, char *argv[])
 	if (cgroup_fd < 0) {
 		fprintf(stderr, "failed to open cgroup path: %s\n",
 			bmc.cgroup_path);
-		return -1;
+		goto out;
 	}
 
 	for (i = 0; i < ARRAY_SIZE(map_infos); i++) {
@@ -696,14 +697,14 @@ static int do_prog_unload(int argc, char *argv[])
 
 	if (!interface_map_path) {
 		fprintf(stderr, "no interface map found\n");
-		return -1;
+		goto out;
 	}
 
 	map_fd = bpf_obj_get(interface_map_path);
 	if (map_fd < 0) {
 		fprintf(stderr, "failed to get map from %s\n",
 			interface_map_path);
-		return -1;
+		goto out;
 	}
 
 	key = 0;
@@ -711,7 +712,7 @@ static int do_prog_unload(int argc, char *argv[])
 	close(map_fd);
 	if (err) {
 		fprintf(stderr, "lookup interface failed\n");
-		return -1;
+		goto out;
 	}
 
 	for (i = 0; i < ARRAY_SIZE(prog_infos); i++) {
@@ -731,7 +732,13 @@ static int do_prog_unload(int argc, char *argv[])
 	for (i = 0; i < ARRAY_SIZE(map_infos); i++)
 		unlink(map_infos[i].pin_path);
 
-	return 0;
+	err = 0;
+out:
+	if (prog_fd != -1)
+		close(prog_fd);
+	if (cgroup_fd != -1)
+		close(cgroup_fd);
+	return err;
 }
 
 static void do_prog_help(void)
-- 
2.17.1

